cmd/gen-key/ecdsa.go

```diff
@@ -13,22 +13,47 @@ import (
 	"github.com/miekg/pkcs11"
 )
 
-var stringToCurve = map[string]elliptic.Curve{
-	"P-224": elliptic.P224(),
-	"P-256": elliptic.P256(),
-	"P-384": elliptic.P384(),
-	"P-521": elliptic.P521(),
+var stringToCurve = map[string]*elliptic.CurveParams{
+	elliptic.P224().Params().Name: elliptic.P224().Params(),
+	elliptic.P256().Params().Name: elliptic.P256().Params(),
+	elliptic.P384().Params().Name: elliptic.P384().Params(),
+	elliptic.P521().Params().Name: elliptic.P521().Params(),
 }
 
-var curveToOID = map[elliptic.Curve][]byte{
-	elliptic.P224(): []byte{6, 5, 43, 129, 4, 0, 33},
-	elliptic.P256(): []byte{6, 8, 42, 134, 72, 206, 61, 3, 1, 7},
-	elliptic.P384(): []byte{6, 5, 43, 129, 4, 0, 34},
-	elliptic.P521(): []byte{6, 5, 43, 129, 4, 0, 35},
+var curveToOIDDER = map[*elliptic.CurveParams][]byte{
+	elliptic.P224().Params(): []byte{6, 5, 43, 129, 4, 0, 33},
+	elliptic.P256().Params(): []byte{6, 8, 42, 134, 72, 206, 61, 3, 1, 7},
+	elliptic.P384().Params(): []byte{6, 5, 43, 129, 4, 0, 34},
+	elliptic.P521().Params(): []byte{6, 5, 43, 129, 4, 0, 35},
 }
 
-func ecArgs(label string, curve elliptic.Curve, compatMode bool) ([]*pkcs11.Mechanism, []*pkcs11.Attribute, []*pkcs11.Attribute) {
-	encodedCurve := curveToOID[curve]
+var oidDERToCurve = map[string]*elliptic.CurveParams{
+	"06052B81040021":       elliptic.P224().Params(),
+	"06082A8648CE3D030107": elliptic.P256().Params(),
+	"06052B81040022":       elliptic.P384().Params(),
+	"06052B81040023":       elliptic.P521().Params(),
+}
+
+var curveToHash = map[*elliptic.CurveParams]crypto.Hash{
+	elliptic.P224().Params(): crypto.SHA256,
+	elliptic.P256().Params(): crypto.SHA256,
+	elliptic.P384().Params(): crypto.SHA384,
+	elliptic.P521().Params(): crypto.SHA512,
+}
+
+var hashToString = map[crypto.Hash]string{
+	crypto.SHA256: "SHA-256",
+	crypto.SHA384: "SHA-384",
+	crypto.SHA512: "SHA-512",
+}
+
+// ecArgs constructs the private and public key template attributes sent to the
```

Comment:
This documentation should specify which return value is for the public vs the private key, or even better, return a `struct { privateKeyAttrs, publicKeyAttrs []*pkcs11.Attribute }`. Same with `rsaArgs` below.

Reply:
Used a special return type.